In [10]:
!ls datalab/

In [11]:
# 查看个人永久空间文件  list files in your permanent storage
!ls /home/tianchi/myspace/

test.py


In [1]:
# 查看当前kernel下已安装的包  list packages
!pip list --format=columns

Package                            Version  
---------------------------------- ---------
absl-py                            0.8.1    
alabaster                          0.7.12   
anaconda-client                    1.7.2    
anaconda-navigator                 1.9.7    
anaconda-project                   0.8.3    
asn1crypto                         0.24.0   
astor                              0.8.0    
astroid                            2.2.5    
astropy                            3.2.1    
atomicwrites                       1.3.0    
attrs                              19.1.0   
Babel                              2.7.0    
backcall                           0.1.0    
backports.functools-lru-cache      1.5      
backports.os                       0.1.1    
backports.shutil-get-terminal-size 1.0.0    
backports.tempfile                 1.0      
backports.weakref                  1.0.post1
beautifulsoup4                     4.7.1    
bitarray                           0.9.3    
bkcharts  

# 第1章 机器学习简介

## 1.1 什么是机器学习

### 机器学习的应用场景
+ 垃圾邮件识别
+ 图像识别
+ 人脸识别
+ 语音识别
+ 数字识别
+ 银行发卡和贷款
+ 搜索推荐
+ 商品推荐
+ 金融预测
+ 医疗诊断
+ 市场分析
+ 无人驾驶
+ 安全领域
+ 自然语言
+ 矿产勘查
+ 宇宙探索
+ 药物研发


## 1.2 课程涵盖的内容和理念

### 本课程会成体系地介绍如下的机器学习算法

+ kNN
+ 线性回归
+ 多项式回归
+ 逻辑回归
+ 模型正则化
+ PCA
+ SVM
+ 决策树
+ 随机森林
+ 集成学习
+ 模型选择
+ 模型调试

其中最常用的算法是：逻辑回归、决策树和随机森林

### 课程的讲解方式

+ 深入理解算法基本原理
+ 实际使用算法解决真实场景的问题
+ 对不同算法进行对比试验
+ 对同一算法的不同参数进行对比试验
+ 对部分算法底层编写

### 介绍如何使用算法
+ 如何评价算法的好坏
+ 如何解决过拟合和欠拟合
+ 如何调节算法的参数
+ 如何验证算法的正确性

### 不仅仅是调库
+ 不反对调库。但是在调库的时候应该对概念原理了解
+ 深入代码内部，可以帮助我们更好地理解算法
+ 更好地理解算法，可以帮助我们更好地选择算法
+ 甚至在将来创造新的算法

### 这门课希望兼顾
+ 算法原理的学习
+ 部分算法底层的编写
+ scikit-learn机器学习库的使用

## 1.3 课程所使用的技术栈

### 环境

+ 语言：Python3
+ 框架：Scikit-learn
+ 其他：numpy and matplotlib
+ IDE : jupyter notebook and PyCharm

### 课程使用的数据集

> sklearn内置数据集或者通过sklearn可以直接下载的数据集


# 第2章 机器学习基础

## 2.1 机器学习世界的数据

+ 数据整体叫数据集(dataset)
+ 每一行数据称为一个样本
+ 除最后一列，每一列表达样本的一个特征(feature)
+ 最后一列，称为标记(label)

### 机器学习的基本概念如下
> ![机器学习的基本概念图示](./images/机器学习的基本概念.jpg)

### 特征向量的表示如下

> ![特征向量的表示](./images/特征向量的表示.jpg)

### 特征空间与分类

> ![特征空间与分类](./images/特征空间与分类.jpg)

### 特征可以很抽象

> ![特征可以很抽象](./images/特征可以很抽象.jpg)

## 2.2 机器学习的主要任务

### 机器学习的任务分类

+ 1.分类任务，只能有几种确定的结果：
  + 1.1 二分类
    + 判断是否是垃圾邮件
    + 判断发卡有没有风险
    + 判断病患是良性肿瘤还是恶性肿瘤
    + 判断股票是涨还是跌
  + 1.2 多分类
    + 数字识别
    + 图像识别
    + 银行判断发卡的风险评级
    + 玩2048游戏，上移下移左移右移
    + 多分类任务
      + 可以转化为二分类任务
      + 只支持二分类的算法可以用来处理多分类任务
      + 有些算法天然可以完成多分类任务
  + 1.3 多标签分类：一个对象可以划分到多个分类中，最后根据这些分类综合区判断对象
+ 2.回归任务：结果是一个**连续数字**的值，而不是类别
  + 房屋价格
  + 市场分析
  + 学生成绩
  + 股票价格

> 有些算法只能用在分类，有的只能回归，有的都行。`一些情况下回归任务可以转化成分类任务`

### 机器学习解决问题的通用套路

![机器学习解决问题的通用套路](images/机器学习解决问题的通用套路.jpg)

## 2-3 监督学习、非监督学习、半监督学习和增强学习

### 监督学习
> 给机器的训练数据拥有“标记”或者“答案”，数据都有标记那一列，除了这一列，其他列表示特征。每一行数据表示一个样本，例如猫狗分类，要告诉机器是猫还是狗。

监督学习举例：
+ 图像已经拥有了标定信息
+ 银行已经积累了一定客户信息和他们信用卡的信用情况
+ 医院已经积累了一定的病人信息和他们最终确诊是否患病的情况
+ 市场积累了房屋的基本信息和最终成交的金额

监督学习主要处理两大类问题:**分类和回归**

监督学习涉及的算法：
+ K近邻
+ 线性回归和多项式回归
+ 逻辑回归
+ SVM
+ 决策树和随机森林

很多时候非监督学习的核心是辅助监督学习

### 非监督学习

> 给机器的训练数据没有任何“标记”或者“答案”

非监督学习的意义：**对没有标记的数据进行分类--聚类分析**

例如：电商网站给消费群体分类

+ **另外一个意义**：对数据进行降维处理
+ **特征提取**：信用卡的信用评级和人的胖瘦无关？这时候可以扔掉人的胖瘦数据，进而达到降维目的
+ **特征压缩**：不扔掉数据，两组数据有很强的关联性，让二维的点变成一维的点，特这个能压缩使用的主要手段是PCA
+ **降维处理的意义**：方便可视化

非监督学习还可以进行异常检测

### 半监督学习

> 一部分数据有“标记”或者“答案”，另一部分数据没有

更常见：**各种原因产生的标记缺失**

通常`先使用非监督学习的手段对数据进行处理，之后使用监督学习手段做模型的训练和预测`

### 增强学习
> 根据周围环境的情况，采取行动，根据采取行动的结果，学习行动方式，类似于PID循环

**应用**：阿尔法狗，无人驾驶，机器人

监督学习和半监督学习依然是基础